### PR TITLE
User management backend

### PR DESCRIPTION
**Pull request by ehrhorn**
**SEE THE END FOR HOW TO USE REFRESH TOKEN**

This PR adds user management in the backend.

Currently you can:
* Sign up via `/signup`; needs `email` and `password`
* Delete user via `/delete`; needs an `email`
* Update user auth level via `/update`; needs an `email` and and `auth_level` (integer between, and including, 1 and 3)
* List all users (and their `auth_level`) via `/list`; needs no arguments, just a `GET`
* NEW: sign in and get a JWT
* NEW: only access certain endpoints with a JWT (see OpenAPI docs for which)

You can see further documentation at our OpenAPI docs.

Working on:
* ~Salting~
* ~JSON web token~
* ~Refresh token cookie~
* ~Logout~
* ~Any exceptions I have forgotten or not yet implemented~
* ~Tests~
* ~Other clean up stuff~

**PLEASE NOTE THE FOLLOWING:**
You must have a `SECRET_KEY` environment variable used by `passlib` for hashing the passwords.
You can create it by running `openssl rand -hex 32` in your terminal.

You _may_ have to delete all your Docker volumes and images... I had to.

Pytest fails because of a too low coverage.

**PLEASE ALSO NOTE:**
`api/public/__init__.py` sets the protected endpoints. I have thus separated out `login` and `signup`, as they do not require login to access. A new route `/user/` is thus created. ~I have moved `login` and `signup` to root, as OpenAPI does not seem to like the `login` endpoint being at e.g. `/auth/login`.~ This last point has been fixed, and `login` is now again at `/auth/login`.

**REFRESH TOKEN HELP**
~At each login you will now receive an access token and a refresh token.~
~The access token has a short expiry time, the refresh token a long one; these are set in `api/config.py:AuthSettings`.~
~The client must store the refresh token in an httponly cookie, and the access token in local storage, as per https://stackoverflow.com/questions/27726066/jwt-refresh-token-flow.~

~The endpoint `/auth/login` takes a form request.~
~As I understand it, this is part of the OAuth spec.~
~Assuming the user `user` with the password `password` exists, you can call the endpoint with HTTPie as follows for login:~

~http --form POST localhost:8000/auth/login username="user" password="password"~

~This'll return to you a new access token and a new refresh token, assuming login is successful.~

~If you wish to use a refresh token for login, you'll instead do:~

~http --form POST localhost:8000/auth/login refresh_token="<your refresh token here>" grant_type="refresh_token"~

~Assuming the refresh token is valid, this'll return an access token and a refresh token as well.~

It is sort of unclear to me what the client exactly wants.
The client now has access to the following:

* An `/auth/login` endpoint that takes a form with `username` and `password`. This yields a JWT and sets an HttpOnly cookie named `terastore_refresh_token`.
* An `/auth/refresh` endpoint that looks for an HttpOnly cooke named `terastore_refresh_token`. If it is found, a new `terastore_refresh_token` HttpOnly cookie is set, and a JWT returned.

The unclear bit is whether the client would only want one endpoint. I have seen someone mention that that is the spec.
However, it'll take a bit longer to implement.

### CODE REVIEW

File: backend/api/utils/exceptions.py [hunk 1]
```diff
@@ -75,3 +75,27 @@ def __init__(
...
+class UserAlreadyExistsError(Exception):
+    """Exception raised when a username already exists."""
+
+    def __init__(
+        self: Self,
+        username: str,
+    ) -> None:
+        self.username = username
+        super().__init__(
+            f"Username {username} already exists.",
```
Comment (Duckling92) [inline backend/api/utils/exceptions.py:100]: I think this should say "email" already exist to not confuse someone trying to create a new user. Maybe all "username" variables should be changed to email as well for our sanity?

File: frontend/src/components/AdminDashboard.tsx [hunk 2]
```diff
@@ -1,4 +1,136 @@
...
+			users.findIndex((u) => u.email === user.email),
+			user,
+		);
+	};
+
+	const deleteUser = (user: BackendUser) => {
+		deleteUserOnBackend(user).then((res) => {
+			if (res.status === 200) {
+				notifications.show({
+					title: "Download pulses",
```
Comment (ehrhorn) [inline frontend/src/components/AdminDashboard.tsx:33]: Is it just me misunderstading? What does "Download pulses" have to do with "User deleted"?

File: frontend/src/components/AdminDashboard.tsx [hunk 3]
```diff
@@ -1,4 +1,136 @@
...
+		if (newAuthLevel === null) return;
+
+		const updatedUser = { ...user };
+		updatedUser.auth_level =
+			BackendAuthLevel[newAuthLevel as keyof typeof BackendAuthLevel];
+		updateUserOnBackend(updatedUser).then((res) => {
+			if (res.status === 200) {
+				notifications.show({
+					title: "Download pulses",
+					message: "User updates.",
```
Comment (ehrhorn) [inline frontend/src/components/AdminDashboard.tsx:98]: Here again, "Download pulses" together with "User updates". Must be something I'm misunderstanding...

File: frontend/src/components/AdminDashboard.tsx (hunk 2 above)
Comment (Duckling92) [reply frontend/src/components/AdminDashboard.tsx:33]: fixed

File: frontend/src/components/AdminDashboard.tsx (hunk 3 above)
Comment (Duckling92) [reply frontend/src/components/AdminDashboard.tsx:98]: You never pull a copy pasta? Fixed.

**Comment by Duckling92** [comment]
Code looks clean.

Before approval, I will ensure CORS works when deployed on VM.

**Comment by Duckling92** [comment]
Fixed. Closes #105

**Review by Duckling92 — APPROVED** [review]

**Pull request merged**